### PR TITLE
Make dry matter loss method for fermentation cumulative instead of daily

### PR DESCRIPTION
**Pull request by Isgotkowitz**
This PR reworks the method for calculating dry matter lost to fermentation so that it can process the degradations from many days at a time, instead of it needing to be run every single simulated day.

### Context

Issue(s) closed by this pull request: addresses #1375

- [x] The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated.
      
How big are the changes in the PR? Would you nominate it for a major version upgrade?
- [ ] Yes
- [x] No

### What

Before the current changes, the method for calculating the dry matter lost to fermentation (which is shared by Silage and Baleage, and implemented in the Storage class) would only calculate the dry matter lost on a single day - now it calculates all dry matter lost over a series of days.

### Why

This is the design for processing degradations that was planned in the Feed Storage design doc. It allows for much quicker processing of degradations over long periods of time because it doesn't require the method to be called every single simulated day, removing a lot of processing overhead.

### How

The method `calculate_dry_matter_loss_to_gas` has been modified to iterate over the weather conditions of many days at a time. While iterating over the weather conditions, a tally of total dry matter lost is kept and the dry matter fraction remaining in the crop is updated and used to calculate degradations on the next iteration.

After processing the degradations, the crop is updated to record the last day it had its degradations processed - this is recorded in the attribute `last_time_degraded` in `HarvestedCrop`.

A helper method, `_get_conditions` has been introduced which handles getting the right series of weather conditions based on the current time and the last time degradations were processed for a crop.

### Test plan

All unit tests for `Storage` and `HarvestedCrop` have been updated to test the new functionality. Also, a `CropManagement` test that broke has had a line added to account for `HarvestedCrop`'s new attribute.

### CODE REVIEW

**Review by KevinP-B — APPROVED** [review]
Elegant solution to problems encountered in the temporal variability of stored feed changes. Not all changes can be effectively calculated daily, nor need they be. This PR reduces computational load and reframes degradations in a flexible, scalable way. Bravo!

File: RUFAS/routines/feed_storage/harvested_crop.py [hunk 1]
```diff
@@ -16,9 +17,12 @@ class HarvestedCrop:
     type : CropType
         The type of the crop (enum), a subdivision of crop category.
     harvest_time : Time
-         The time at which the crop was harvested
+        The time at which the crop was harvested
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/harvested_crop.py]:
```suggestion
        The time at which the crop was harvested.
```

File: RUFAS/routines/feed_storage/harvested_crop.py [hunk 2]
```diff
@@ -16,9 +17,12 @@ class HarvestedCrop:
...
     storage_time : Time
-         The time at which the crop was stored
+        The time at which the crop was stored
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/harvested_crop.py]:
```suggestion
        The time at which the crop was stored.
```

Comment (ew3361zh) [reply RUFAS/routines/feed_storage/harvested_crop.py]: What do you think about renaming this to something like `storage_start_time`? I think `storage_time` sounds more like a period of time over which the crop was stored.

File: RUFAS/routines/feed_storage/storage.py [hunk 3]
```diff
@@ -215,65 +214,87 @@ def record_stored_crops(self, gaseous_dry_matter_loss: float) -> None:
         """
         pass
 
-    def calculate_dry_matter_loss_to_gas(
-        self, crop: HarvestedCrop, current_conditions: CurrentDayConditions, time: Time
-    ) -> float:
+    def calculate_dry_matter_loss_to_gas(self, crop: HarvestedCrop, conditions: list[CurrentDayConditions]) -> float:
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/storage.py]:
```suggestion
    def calculate_dry_matter_loss_to_gas(self, crop: HarvestedCrop, weather_conditions: list[CurrentDayConditions]) -> float:
```

File: RUFAS/routines/feed_storage/storage.py [hunk 4]
```diff
@@ -215,65 +214,87 @@ def record_stored_crops(self, gaseous_dry_matter_loss: float) -> None:
...
 
         Parameters
         ----------
         crop : HarvestedCrop
             The stored crop that is losing dry matter.
-        current_conditions : CurrentDayConditions
-            Current conditions of the simulated day.
-        time : Time
-            Time instance containing the current time of the simulation.
+        conditions : list[CurrentDayConditions]
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/storage.py]:
```suggestion
        weather_conditions : list[CurrentDayConditions]
```

File: RUFAS/routines/feed_storage/storage.py [hunk 5]
```diff
@@ -215,65 +214,87 @@ def record_stored_crops(self, gaseous_dry_matter_loss: float) -> None:
...
+            List of daily conditions over which dry matter loss will be calculated for.
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/storage.py]:
```suggestion
            List of daily weather conditions over which dry matter loss will be calculated.
```

File: RUFAS/routines/feed_storage/storage.py [hunk 6]
```diff
@@ -215,65 +214,87 @@ def record_stored_crops(self, gaseous_dry_matter_loss: float) -> None:
...
 
-        if is_alfalfa:
-            dry_matter_loss_fraction = 0.0156 - 0.0364 * (dry_matter_fraction - 0.20)
-        else:
-            dry_matter_loss_fraction = 0.00864 - 0.0193 * (dry_matter_fraction - 0.15)
+    def _get_conditions(
+        self, last_degradations_time: Time, current_time: Time, weather: Weather
+    ) -> list[CurrentDayConditions]:
+        """
+        Gets a series of weather conditions for a time period which will have weather conditions calculated for it.
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/storage.py]: This description is a little confusing for me. I made a suggestion but not sure if it's capturing what's happening still.
```suggestion
        Gets a series of weather conditions for a time period.
```

File: RUFAS/routines/feed_storage/storage.py [hunk 7]
```diff
@@ -215,65 +214,87 @@ def record_stored_crops(self, gaseous_dry_matter_loss: float) -> None:
...
 
-        return crop.dry_matter_mass * dry_matter_loss_fraction
+        Parameters
+        ----------
+        last_degradations_time : Time
+            Time instance recording the last day a crop's degradations were processed.
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/storage.py]:
```suggestion
            Time instance for the last day a crop's degradations were processed.
```

**Review by ew3361zh — APPROVED** [review]
LGTM. Testing suite was passing as expected.
I made some minor suggestions for clarity/grammar. Nice job, Ed!

File: RUFAS/routines/feed_storage/storage.py (hunk 6 above)
Comment (Isgotkowitz) [reply RUFAS/routines/feed_storage/storage.py]: Good catch, I don't even know what I meant.

**Pull request merged**